src/core/components/pin/gc.js

```diff
@@ -0,0 +1,153 @@
+      if (err) {
+        return cb(new Error(`Could not list pinner internal blocks: ${err.message}`))
+      }
+      log(`Found ${cids.length} pinner internal blocks`)
+      // log('  ' + cids.join('\n  '))
+      cb(null, cids)
+    }),
+
+    // The MFS root and all its descendants
+    (cb) => ipfs._repo.root.get(MFS_ROOT_DS_KEY, (err, mh) => {
+      if (err) {
+        if (err.code === 'ERR_NOT_FOUND') {
+          log(`No blocks in MFS`)
+          return cb(null, [])
+        }
+        return cb(new Error(`Could not get MFS root from datastore: ${err.message}`))
+      }
+
+      getDescendants(ipfs, new CID(mh), cb)
+    })
+  ], (err, res) => {
+    if (err) {
+      return callback(err)
+    }
+
+    const cids = [].concat(...res).map(cid => cid.toV1().toString('base32'))
+    return callback(null, new Set(cids))
+  })
+}
+
+// Recursively get descendants of the given CID
+function getDescendants (ipfs, cid, callback) {
+  ipfs.refs(cid, { recursive: true }, (err, refs) => {
+    if (err) {
+      return callback(new Error(`Could not get MFS root descendants from store: ${err.message}`))
+    }
+    const cids = [cid, ...refs.map(r => new CID(r.ref))]
+    log(`Found ${cids.length} MFS blocks`)
+    // log('  ' + cids.join('\n  '))
+    callback(null, cids)
+  })
+}
+
+// Delete all blocks that are not marked as in use
+function deleteUnmarkedBlocks (ipfs, markedSet, blocks, start, callback) {
+  // Iterate through all blocks and find those that are not in the marked set
+  // The blocks variable has the form { { key: Key() }, { key: Key() }, ... }
+  const unreferenced = []
+  const res = []
+  let errCount = 0
+  for (const { key: k } of blocks) {
+    try {
+      const cid = dsKeyToCid(k)
+      const b32 = cid.toV1().toString('base32')
+      if (!markedSet.has(b32)) {
+        unreferenced.push(cid)
+      }
+    } catch (err) {
+      errCount++
+      const msg = `Could not convert block with key '${k}' to CID: ${err.message}`
+      log(msg)
+      res.push({ err: new Error(msg) })
+    }
+  }
+
+  const msg = `Marked set has ${markedSet.size} unique blocks. Blockstore has ${blocks.length} blocks. ` +
+    `Deleting ${unreferenced.length} blocks.` + (errCount ? ` (${errCount} errors)` : '')
+  log(msg)
+  // log('  ' + unreferenced.join('\n  '))
+
+  mapLimit(unreferenced, BLOCK_RM_CONCURRENCY, (cid, cb) => {
+    // Delete blocks from blockstore
+    ipfs._repo.blocks.delete(cid, (err) => {
+      const res = {
+        cid: cid.toString(),
+        err: err && new Error(`Could not delete block with CID ${cid}: ${err.message}`)
+      }
+      cb(null, res)
+    })
+  }, (_, delRes) => {
+    log(`Complete (${Date.now() - start}ms)`)
+
+    callback(null, res.concat(delRes))
+  })
+}
+
+function dsKeyToCid (key) {
+  // Block key is of the form /<base32 encoded string>
+  const decoder = new base32.Decoder()
+  const buff = decoder.write(key.toString().slice(1)).finalize()
+  return new CID(Buffer.from(buff))
+}
```

Comment:
Would be nice if this util was exported from repo.

Reply:
That's true, @vasco-santos did this in https://github.com/ipfs/js-ipfs/pull/2184/files#diff-47300e7ecd8989b6246221de88fc9a3cR142

Reply:
As we eventually want to move the repo codebase to this repo, I think we should have it here. What do you think?

Reply:
Yes I think that makes sense, I'll add it

Reply:
Actually sorry I'm just noticing that @alanshaw is suggesting this be in repo, which I think makes more sense